nodestream/pipeline/transformers/transformer.py

```diff
@@ -54,83 +57,64 @@ def __init__(
-                # In order to fully respect the Flush,
-                # we need to wait for all pending tasks to complete.
-                # If we're still waiting we can yield the processor.
-                while pending_tasks:
-                    async for result in drain_completed_tasks_async():
-                        yield result
-                yield record
+    async def drain_completed_tasks(self):
+        tasks_drained = 0
+        remaining_pending_tasks = []
+        for task in self.pending_tasks:
+            if task.done():
+                yield task.result()
+                tasks_drained += 1
             else:
-                # Submit the work to the thread pool (only if we have capacity)
-                # If we don't have capacity, yield completed tasks.
-                # Once there is nothing remaining to yield from these tasks, yield the processor until capacity is available.
-                submitted = False
-                while not submitted:
-                    if len(pending_tasks) < self.maximum_pending_tasks:
-                        task = self.thread_pool.submit(self.do_work_on_record, record)
-                        pending_tasks.append(task)
-                        submitted = True
-                    async for result in drain_completed_tasks_async():
-                        yield result
-
-        # After we've finished enqueuing records, we need to drain all tasks.
-        # Items yielded from this loop are the final results of the step.
-        self.logger.debug("Finished enqueuing records, draining pending tasks")
-        while pending_tasks:
-            async for result in drain_completed_tasks_async():
-                yield result
+                remaining_pending_tasks.append(task)
+
+        self.pending_tasks[:] = remaining_pending_tasks
+        if tasks_drained:
+            self.logger.debug(
+                "Drained %s completed tasks, %s pending tasks remain",
+                tasks_drained,
+                len(self.pending_tasks),
+            )
+
+    async def process_record(self, record: Any, _) -> Any:
+        if record is Flush:
+            async for outstanding in self.emit_outstanding_records():
+                yield outstanding
+            yield record
+        else:
+            # Submit the work to the thread pool (only if we have capacity)
+            # If we don't have capacity, yield completed tasks.
+            # Once there is nothing remaining to yield from these tasks,
+            # yield the processor until capacity is available.
+            submitted = False
+            while not submitted:
+                if len(self.pending_tasks) < self.maximum_pending_tasks:
+                    task = self.thread_pool.submit(self.do_work_on_record, record)
+                    self.pending_tasks.append(task)
+                    submitted = True
+                async for result in self.drain_completed_tasks():
```

Comment:
I think using something like `async def emit_outstanding_records(self): ` here is better since drain_completed_tasks() does not yield the processor at any point. Unless this piece is intended on blocking (so something like the extractor does not put more records with the new channel update.) this should yield the processor so the pipeline doesn't get stuck waiting for results in this blocking section.

Reply:
Hmm yes... I see what you mean. this is mend to be calling `emit_outstanding_records` which does yield.